File: dsa-problems/SegmentTree/solutions/cpp/SEG-009-range-t-threshold-majority.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <string>
#include <array>

using namespace std;

struct Summary {
    vector<pair<int, int>> candidates;
    static const int K = 40;
    
    void add(int val, int count) {
        for (auto& p : candidates) {
            if (p.first == val) {
                p.second += count;
                return;
            }
        }
        candidates.push_back({val, count});
        if (candidates.size() > K) {
            int minCnt = 2e9;
            for (auto& p : candidates) minCnt = min(minCnt, p.second);
            
            vector<pair<int, int>> next;
            for (auto& p : candidates) {
                p.second -= minCnt;
                if (p.second > 0) next.push_back(p);
            }
            candidates = next;
        }
    }
    
    void merge(const Summary& other) {
        for (const auto& p : other.candidates) {
            add(p.first, p.second);
        }
    }
};

class Solution {
    vector<Summary> tree;
    vector<vector<int>> positions;
    vector<int> idToVal;
    int n;

    void build(const vector<int>& a, int node, int start, int end) {
        if (start == end) {
            tree[node].add(a[start], 1);
        } else {
            int mid = (start + end) / 2;
            build(a, 2 * node + 1, start, mid);
            build(a, 2 * node + 2, mid + 1, end);
            
            tree[node] = Summary(); // Reset
            tree[node].merge(tree[2 * node + 1]);
            tree[node].merge(tree[2 * node + 2]);
        }
    }

    Summary query(int node, int start, int end, int l, int r) {
        if (l > end || r < start) return Summary();
        if (l <= start && end <= r) return tree[node];
        
        int mid = (start + end) / 2;
        Summary s1 = query(2 * node + 1, start, mid, l, r);
        Summary s2 = query(2 * node + 2, mid + 1, end, l, r);
        
        s1.merge(s2);
        return s1;
    }

    int getFreq(int valId, int l, int r) {
        const auto& pos = positions[valId];
        auto it1 = lower_bound(pos.begin(), pos.end(), l);
        auto it2 = upper_bound(pos.begin(), pos.end(), r);
        return distance(it1, it2);
    }

public:
    vector<int> process(const vector<int>& arr, const vector<array<int,3>>& queries) {
        n = arr.size();
        
        // Coordinate Compression
        map<int, int> valToId;
        int idCounter = 0;
        vector<int> mappedArr(n);
        
        for (int x : arr) {
            if (valToId.find(x) == valToId.end()) {
                valToId[x] = idCounter++;
                idToVal.push_back(x);
                positions.push_back({});
            }
        }
        for (int i = 0; i < n; i++) {
            mappedArr[i] = valToId[arr[i]];
            positions[mappedArr[i]].push_back(i);
        }
        
        tree.assign(4 * n, Summary());
        build(mappedArr, 0, 0, n - 1);
        
        vector<int> results;
        for (const auto& q : queries) {
            int l = q[0];
            int r = q[1];
            int t = q[2];
            
            Summary s = query(0, 0, n - 1, l, r);
            vector<int> cands;
            for (auto& p : s.candidates) cands.push_back(p.first);
            for (int i = 0; i < 40; i++) cands.push_back(mappedArr[l + rand() % (r - l + 1)]);
            sort(cands.begin(), cands.end());
            cands.erase(unique(cands.begin(), cands.end()), cands.end());
            
            int bestVal = -1;
            int maxFreq = -1;
            
            for (int valId : cands) {
                int freq = getFreq(valId, l, r);
                if (freq >= t) {
                    int realVal = idToVal[valId];
                    if (freq > maxFreq) {
                        maxFreq = freq;
                        bestVal = realVal;
                    } else if (freq == maxFreq) {
                        if (bestVal == -1 || realVal < bestVal) {
                            bestVal = realVal;
                        }
                    }
                }
            }
            results.push_back(bestVal);
        }
        return results;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, q;
    if (!(cin >> n >> q)) return 0;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];
    vector<array<int, 3>> queries(q);
    for (int i = 0; i < q; i++) {
        string type;
        cin >> type; // MAJ
        cin >> queries[i][0] >> queries[i][1] >> queries[i][2];
    }
    Solution sol;
    vector<int> results = sol.process(arr, queries);
    for (int res : results) {
        cout << res << "\n";
    }
    return 0;
}

File: dsa-problems/SegmentTree/solutions/java/SEG-009-range-t-threshold-majority.java
import java.util.*;

class Solution {
    private List<Integer>[] positions;
    private int[] arr;
    
    // Misra-Gries summary size
    private static final int K = 40; 
    
    static class Summary {
        // value -> count
        // We use a simple list of pairs for small K
        List<int[]> candidates = new ArrayList<>();
        
        void add(int val, int count) {
            for (int[] c : candidates) {
                if (c[0] == val) {
                    c[1] += count;
                    return;
                }
            }
            candidates.add(new int[]{val, count});
            if (candidates.size() > K) {
                // Prune
                // Find min count
                int minCnt = Integer.MAX_VALUE;
                for (int[] c : candidates) minCnt = Math.min(minCnt, c[1]);
                
                List<int[]> next = new ArrayList<>();
                for (int[] c : candidates) {
                    c[1] -= minCnt;
                    if (c[1] > 0) next.add(c);
                }
                candidates = next;
            }
        }
        
        void merge(Summary other) {
            for (int[] c : other.candidates) {
                add(c[0], c[1]);
            }
        }
    }
    
    private Random random = new Random();
    private Summary[] tree;
    private int n;

    public int[] process(int[] arr, int[][] queries) {
        this.arr = arr;
        this.n = arr.length;
        
        // Coordinate Compression / Positions Map
        Map<Integer, Integer> valToId = new HashMap<>();
        List<Integer> idToVal = new ArrayList<>();
        int idCounter = 0;
        
        for (int x : arr) {
            if (!valToId.containsKey(x)) {
                valToId.put(x, idCounter++);
                idToVal.add(x);
            }
        }
        
        positions = new List[idCounter];
        for (int i = 0; i < idCounter; i++) positions[i] = new ArrayList<>();
        
        int[] mappedArr = new int[n];
        for (int i = 0; i < n; i++) {
            mappedArr[i] = valToId.get(arr[i]);
            positions[mappedArr[i]].add(i);
        }
        
        // Build Segment Tree
        tree = new Summary[4 * n];
        build(mappedArr, 0, 0, n - 1);
        
        int[] results = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int l = queries[i][0];
            int r = queries[i][1];
            int t = queries[i][2];
            
            Summary s = query(0, 0, n - 1, l, r);
            Set<Integer> candidates = new HashSet<>();
            for (int[] c : s.candidates) candidates.add(c[0]);
            
            // Random sampling
            for (int k = 0; k < 40; k++) {
                int randIdx = l + random.nextInt(r - l + 1);
                candidates.add(mappedArr[randIdx]);
            }
            
            int bestVal = -1;
            int maxFreq = -1;
            
            for (int valId : candidates) {
                int freq = getFreq(valId, l, r);
                if (freq >= t) {
                    int realVal = idToVal.get(valId);
                    if (freq > maxFreq) {
                        maxFreq = freq;
                        bestVal = realVal;
                    } else if (freq == maxFreq) {
                        if (bestVal == -1 || realVal < bestVal) {
                            bestVal = realVal;
                        }
                    }
                }
            }
            results[i] = bestVal;
        }
        return results;
    }
    
    private int getFreq(int valId, int l, int r) {
        List<Integer> pos = positions[valId];
        int leftIdx = Collections.binarySearch(pos, l);
        if (leftIdx < 0) leftIdx = -leftIdx - 1;
        int rightIdx = Collections.binarySearch(pos, r);
        if (rightIdx < 0) rightIdx = -rightIdx - 2;
        
        if (leftIdx > rightIdx) return 0;
        return rightIdx - leftIdx + 1;
    }

    private void build(int[] a, int node, int start, int end) {
        if (start == end) {
            tree[node] = new Summary();
            tree[node].add(a[start], 1);
        } else {
            int mid = (start + end) / 2;
            build(a, 2 * node + 1, start, mid);
            build(a, 2 * node + 2, mid + 1, end);
            tree[node] = new Summary();
            tree[node].merge(tree[2 * node + 1]);
            tree[node].merge(tree[2 * node + 2]);
        }
    }

    private Summary query(int node, int start, int end, int l, int r) {
        if (l > end || r < start) return new Summary();
        if (l <= start && end <= r) return tree[node];
        
        int mid = (start + end) / 2;
        Summary s1 = query(2 * node + 1, start, mid, l, r);
        Summary s2 = query(2 * node + 2, mid + 1, end, l, r);
        
        Summary res = new Summary();
        res.merge(s1);
        res.merge(s2);
        return res;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (sc.hasNextInt()) {
            int n = sc.nextInt();
            int q = sc.nextInt();
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) arr[i] = sc.nextInt();
            int[][] queries = new int[q][3];
            for (int i = 0; i < q; i++) {
                String type = sc.next(); // MAJ
                queries[i][0] = sc.nextInt();
                queries[i][1] = sc.nextInt();
                queries[i][2] = sc.nextInt();
            }
            Solution sol = new Solution();
            int[] results = sol.process(arr, queries);
            for (int res : results) {
                System.out.println(res);
            }
        }
        sc.close();
    }
}
